src/ir/auth_logic/ast_construction.cc

```diff
@@ -90,14 +102,53 @@ static Fact ConstructFact(AuthLogicParser::FactContext& context) {
   return ConstructFact(*CHECK_NOTNULL(ctx.fact()));
 }
 
+static Assertion ConstructAssertion(
+    AuthLogicParser::AssertionContext& context) {
+  if (auto* ctx =
+          dynamic_cast<AuthLogicParser::FactAssertionContext*>(&context)) {
+    return Assertion(ConstructFact(*CHECK_NOTNULL(ctx->fact())));
+  }
+  auto& ctx = *CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::HornClauseAssertionContext*>(&context));
+  std::vector<BaseFact> base_facts = utils::MapIter<BaseFact>(
+      ctx.rvalue(), [](AuthLogicParser::RvalueContext* rvalue_context) {
+        return ConstructRvalue(*CHECK_NOTNULL(rvalue_context));
+      });
+  return Assertion(ConditionalAssertion(
+      ConstructFact(*CHECK_NOTNULL(ctx.fact())), std::move(base_facts)));
+}
+
+static SaysAssertion ConstructSaysAssertion(
+    AuthLogicParser::SaysAssertionContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::SaysSingleContext*>(&context)) {
```

Comment:
Don't abbreviate unnecessarily. Meaningful names makes it easier to read code: `ctx` -> `says_single_context`